src/array/range.ts

```diff
@@ -0,0 +1,48 @@
+/**
+ * Returns an array of numbers from `start` to `end` with the specified `step` value.
+ *
+ * This function generates a range of numbers, starting from `start`, ending before `end`,
+ * and incrementing by `step`. If `step` is not provided, it defaults to `1` for an
+ * ascending range and `-1` for a descending range.
+ *
+ * @param {number} start - The starting number of the range.
+ * @param {number} [end] - The end number of the range.
+ * @param {number} [step] - The step value for the range.
+ * @returns {number[]} An array of numbers from `start` to `end` with the specified `step`.
+ *
+ * @example
+ * // Returns [0, 1, 2, 3]
+ * range(4);
+ *
+ * @example
+ * // Returns [0, 5, 10, 15]
+ * range(0, 20, 5);
+ *
+ * @example
+ * // Returns [0, -1, -2, -3]
+ * range(0, -4, -1);
+ *
+ * @example
+ * // Returns [1, 1, 1]
+ * range(1, 4, 0);
+ */
+
+export function range(start: number, end?: number, step?: number): number[] {
+  if (end === undefined) {
+    end = start;
+    start = 0;
+  }
+
+  if (step === undefined) {
+    step = start <= end ? 1 : -1;
+  }
+
+  const length = Math.max(Math.ceil((end - start) / (step || 1)), 0);
+  const result = new Array(length);
+
+  for (let i = 0; i < length; i++) {
+    result[i] = start + i * step;
+  }
+
+  return result;
```

Comment:
```suggestion
  const result = [];

  for (let i = start; i < end; i += step) {
    result.push(start);
  }
  
  return result;
```

I guess implementation like this might be more readable, what do you think?

Reply:
I agree it's more readable, but each result.push(i) involves dynamic array resizing, which can be costly. Preallocating a fixed array size demonstrates better performance. I tested this yesterday, and the benchmarks showed it was at least twice as fast, if I remember correctly. 
IMHO, given that we're aiming for a high-performance utility library, it makes sense to prioritize performance considerations a bit more.

Reply:
This is the result of the benchmark I measured again:
Preallocating a fixed array size shows 2~3 times faster than `result.push` in for loop. Let me know what do you think!

Array size is 100:
<img width="1129" alt="Screenshot 2024-06-19 at 11 15 27 PM" src="https://github.com/toss/es-toolkit/assets/47740690/913b5186-c6a7-486a-ae70-081cf5e4dacf">

Array size is 10000:
<img width="1063" alt="Screenshot 2024-06-19 at 11 15 45 PM" src="https://github.com/toss/es-toolkit/assets/47740690/7561ebb0-440b-47e1-a44d-30b7fc01b333">

Reply:
If we have this performance difference, I guess we should allocate a fixed-size array in advance.